plugins/cassandra/cassandraembed.cpp

```diff
@@ -4408,37 +5128,55 @@ class CCasssandraWorkUnitFactory : public CWorkUnitFactory, implements ICassandr
         return getUnsignedResult(NULL, getSingleResult(result)) != 0; // Shouldn't be more than 1, either
     }
-
-    IConstWorkUnitIterator * getSortedWorkUnits(ISecManager *secmgr, ISecUser *secuser)
+    void mergeFilter(CIArrayOf<PostFilter> &filters, WUSortField field, const char *value)
     {
-        // Hack in some test code to test stream merging
-        Owned<IPTree> parent = createPTree("WorkUnits");
-        Owned<CassMultiIterator> merger = new CassMultiIterator(0, true); // Merge by wuid (note that we didn't fetch partition...)
-        for (int i = 0; i < NUM_PARTITIONS; i++)
+        // Combine multiple filters on wuid - Cassandra doesn't like seeing more than one.
+        ForEachItemIn(idx, filters)
         {
-            merger->addResult(*new CassandraResult(fetchDataByPartition(workunitInfoMappings, i)));
+            PostFilter &filter = filters.item(idx);
+            if (filter.queryField()==field)
+            {
+                const char *prevLimit = filter.queryValue();
+                int diff = strcmp(prevLimit, value);
+                if (diff && ((diff < 0) == (field==WUSFwuid)))
+                    filter.setValue(value);
+                return;
+            }
         }
-        return createSecureConstWUIterator(merger.getClear(), secmgr, secuser);
+        // Not found - add new filter
+        filters.append(*new PostFilter(field, value, true));
     }
-
     IConstWorkUnitIterator * getWorkUnitsByXXX(const char *xpath, const char *key, ISecManager *secmgr, ISecUser *secuser)
     {
+        Owned<CassMultiIterator> merger = new CassMultiIterator(NULL, 0, 0, true); // Merge by wuid
         if (!key || !*key)
-            return getSortedWorkUnits(secmgr, secuser);
-        const CassandraXmlMapping *mappings = searchMappings+3;  // Don't return the xpath, searchPrefix or searchValue fields
-        Owned<CassandraResult> result = new CassandraResult(fetchDataForKey(xpath, key));
-        Owned<IPTree> parent = createPTree("WorkUnits");
-        CassandraIterator rows(cass_iterator_from_result(*result));
-        while (cass_iterator_next(rows))
         {
-            Owned<IPTree> wuXML = rowToPTree(xpath, key, mappings, cass_iterator_get_row(rows));
-            const char *wuid = wuXML->queryName();
-            parent->addPropTree(wuid, wuXML.getClear());
+            CIArrayOf<PostFilter> wuidFilters;
+            for (int i = 0; i < NUM_PARTITIONS; i++)
+            {
+                merger->addResult(*new CassandraResult(fetchDataByPartition(workunitInfoMappings, i, wuidFilters)));
+            }
+        }
+        else
+            merger->addResult(*new CassandraResult(fetchDataForKey(xpath, key)));
+        return createSecureConstWUIterator(merger.getClear(), secmgr, secuser);
+    }
+    StringArray &_getUniqueValues(const char *xpath, const char *prefix, StringArray &result) const
+    {
+        if (prefix && strlen(prefix) >= CASS_SEARCH_PREFIX_SIZE)
```

Comment:
Not sure - it might be ok to cache when being used to populate dropdowns, but not so sure if it is for searching workunits.